src/main/java/digital/slovensko/autogram/core/eforms/EFormUtils.java

```diff
@@ -0,0 +1,284 @@
+    }
+
+    public static byte[] getTransformation(String url) {
+        return getResource(url + "/form.sb.xslt");
+    }
+
+    public static byte[] getSchema(String url) {
+        return getResource(url + "/form.xsd");
+    }
+
+    private static byte[] getResource(String url) {
+        var offlineFileLoader = new FileCacheDataLoader();
+        offlineFileLoader.setCacheExpirationTime(21600000);
+        offlineFileLoader.setDataLoader(new CommonsDataLoader());
+
+        DSSDocument xsltDoc;
+        try {
+            xsltDoc = offlineFileLoader.getDocument(url);
+        } catch (DSSException e) {
+            return null;
+        }
+
+        if (xsltDoc == null)
+            return null;
+
+        try {
+            return xsltDoc.openStream().readAllBytes();
+        } catch (IOException e) {
+            return null;
+        }
+    }
+
+    public static Document getXmlFromDocument(DSSDocument documentToDisplay) throws XMLValidationException {
+        try {
+            var builderFactory = DocumentBuilderFactory.newInstance();
+            builderFactory.setNamespaceAware(true);
+            builderFactory.setFeature("http://xml.org/sax/features/external-general-entities", false);
+
+            var is = documentToDisplay.openStream();
+            var inputSource = new InputSource(is);
+            inputSource.setEncoding(ENCODING.displayName());
+            var parsedDocument = builderFactory.newDocumentBuilder().parse(inputSource);
+
+            return parsedDocument;
+
+        } catch (Exception e) {
+            throw new XMLValidationException("Zlyhala validácia XML Datacontainera", "Nepodarilo sa načítať XML dokument", e);
+        }
+    }
+
+    public static Document getEformXmlFromXdcDocument(DSSDocument document) throws XMLValidationException {
+        var xmlDocument = getXmlFromDocument(document);
+        var xmlData = xmlDocument.getElementsByTagNameNS("http://data.gov.sk/def/container/xmldatacontainer+xml/1.1", "XMLData").item(0);
+
+        if (xmlData == null)
+            throw new XMLValidationException("Zlyhala validácia XML Datacontainera", "Element XMLData sa nepodarilo nájsť v XML Dataconatineri");
+
+        var builderFactory = DocumentBuilderFactory.newInstance();
+        builderFactory.setNamespaceAware(true);
+
+        Document responseDocument = null;
+        try {
+            builderFactory.setFeature("http://xml.org/sax/features/external-general-entities", false);
+            responseDocument = builderFactory.newDocumentBuilder().newDocument();
+        } catch (ParserConfigurationException e) {
+            throw new UnrecognizedException(e);
+        }
+
+        if (xmlData.getFirstChild() == null)
+            throw new XMLValidationException("Zlyhala validácia XML Datacontainera", "Element XMLData je prázdny");
+
+        var idk = xmlData.getFirstChild();
+        if (idk.getNodeType() == Node.TEXT_NODE) {
+            if (!idk.getNodeValue().matches("\\s*"))
```

Comment:
Huh, co toto kontroluje?

Reply:
To asi stojí aj za komentár. Ak je vstupné xml napr. pekne odsadené, že medzi elementami sú nejaké whitespaces, getFirstChild() vráti tieto whitespaces ako prvý TEXT_NODE. Toto je problém, že pri vyberaní obsahu z XDC, chceme first child, ale chceme ignorvať tento problém, keď tam sú iba whitespaces. Regex `\s` je whitespace. A vtedy chceme až druhý child node.

Reply:
pridal som komentár

Reply:
Nastáva to, ak vstupné XML je naformátované/odsadené, teda nie je iba celé v jednom riadku.

Reply:
A ta "kanonakikanalizacia" toto nema riesit?

Reply:
Tá sa ale deje neskôr pri podpisovaní. Teoreticky vie integrátor poslať takú kanonikalizáciu, ktorá to nevyhodí. Tuto to parsujeme preto, že chceme nájsť identifier, podľa ktorého chceme nájsť xsd a xslt.

Možno to v praxi stačí prehnať cez tú kanonikalizáciu, ale tento regex nie je zas taký komplikovaný, že by to nechtiac pokazil. Súvisí to s tým, čo sme nižšie písali, že by to chcelo celé asi rozparsované niekam odkladať. Tak tam by sa to mohlo rovno aj kanonikalizovať na začiatku.